File: anb/leetcode/array/53. Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxi = -100000000
        ans = 0
        cur = 0
        neg = True
        
        for num in nums:
            neg = neg and num < 0
            
            maxi = max(maxi, num)
            cur = max(num, cur + num)
            ans = max(ans, cur)
        
        if neg:
            return maxi
        else:
            return ans
        

File: gunkim/leetcode/array/53.Maximum Subarray.java
class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.maxSubArray(new int[] { -2, 1, -3, 4, -1, 2, 1, -5, 4 }));
    }
}

class Solution {
    public int maxSubArray(int[] nums) {
        int maxValue = nums[0];
        int prevValue = nums[0];

        for (int i = 1; i < nums.length; i++) {
            if (prevValue + nums[i] > nums[i]) {
                prevValue += nums[i];
            } else {
                prevValue = nums[i];
            }
            if (prevValue > maxValue) {
                maxValue = prevValue;
            }
        }
        return maxValue;
    }
}